---
title: "Cell-Level Differential Expression Analysis"
author: "Zhaoheng Li, Wei Sun"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: journal
    highlight: tango
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: true
      smooth_scroll: false
    number_sections: false
  df_print: paged
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r,warning = F, message= F}
 library(Seurat)
library(ggpubr)
library(viridis)
library(tidyr)
library(dplyr)
library(tidyr)
library(readxl)
library(stringr)
library(gridExtra)
library(grid)
library(purrr)
library(tidyverse)
meta_dir = "/fh/fast/sun_w/zhaoheng_li/MSK_DEC24/Analysis"
data_dir = "/fh/fast/sun_w/zhaoheng_li/MSK_DEC24/Analysis"
out_dir = "/fh/fast/sun_w/zhaoheng_li/MSK_DEC24/Analysis/outs/differential_expression"
celllevel_out_dir = file.path(out_dir,"cell-level")

```



```{r}
meta = read.csv(file.path(meta_dir,"processed_metadata.csv"))
all_genotypes = setdiff(unique(meta$genotype),"WT")
integrated = readRDS(file.path(data_dir,"integrated-rpca_BDF-ref_withMeta.RDS"))

# remove cells without cell type annotations
keep.cells= rownames(integrated@meta.data)[
  !is.na(integrated@meta.data$celltype)
]

integrated= subset(
  integrated,
  cells = keep.cells
)
integrated$celltype = as.factor(integrated$celltype)
summary(integrated$celltype)
```


```{r}
for(tp in c("D-1","D3","D7","D11","D18")){
  if(!file.exists(file.path(celllevel_out_dir,paste0("DE_cell-level_",tp,".csv")))){
    obj = integrated%>%subset(subset = time_point == tp)
    ct2test = names(table(obj$celltype))[table(obj$celltype)>60]
    print(ct2test) 
    
    res_all = NULL
    for(ct in ct2test){
      print(paste0("current celltype: ", ct))
      obj_sub = obj%>%subset(subset = celltype == ct)
      print(table(obj_sub$genotype))
      for(current_KO in all_genotypes){
        print(paste0("working on: ", current_KO))
        n_KO = sum(obj_sub$genotype==current_KO)
        if(n_KO>=30){
          df <- FindMarkers(obj_sub, 
                            ident.1 = current_KO, ident.2 = 'WT',group.by = 'genotype')
          df$time_point = tp
          df$genotype = current_KO
          df$celltype = ct
          df$gene = rownames(df)
          res_all = rbind(res_all, df)
        }
      }
    }
    write.csv(res_all,file.path(celllevel_out_dir,paste0("DE_cell-level_",tp,".csv")))
  }
}



```

# Session information

```{r}
gc()
sessionInfo()
```


